docker/Dockerfile

```diff
@@ -0,0 +1,203 @@
+RUN pip install garage && \
+    pip install garage[all] && \
+    rm -r $HOME/.cache/pip
+#    && \
+#    if [[ -s $HOME/.mujoco/mjkey.txt ]]; then python -c 'import mujoco_py'; fi
+
+COPY --chown=$USER:$USER docker/entrypoint-runtime.sh $HOME/code/garage/docker/entrypoint-runtime.sh
+ENTRYPOINT ["code/garage/docker/entrypoint-runtime.sh"]
+
+
+# Headless machine with xvfb, i.e. so rendering works #################################################################
+FROM garage-18.04 AS garage-headless-18.04
+USER root
+RUN \
+  apt-get -y -q update && \
+  # Prevents debconf from prompting for user input
+  # See https://github.com/phusion/baseimage-docker/issues/58
+  DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    # Dummy X server
+    xvfb \
+    pulseaudio && \
+  apt-get clean && \
+  rm -rf /var/lib/apt/lists/* && \
+  mkdir -p -m 1777 /tmp/.X11-unix
+USER $USER
+
+# Ready, set, go.
+COPY --chown=$USER:$USER docker/entrypoint-headless.sh $HOME/code/garage/docker/entrypoint-headless.sh
+ENTRYPOINT ["code/garage/docker/entrypoint-headless.sh"]
+
+
+# Special setup for garage developers, which installs garage from source ##############################################
+# Assumes $PWD is the garage repository.
+FROM garage-headless-18.04 AS garage-dev-18.04
+
+# Copy over just setup.py first, so the Docker cache doesn't expire until
+# dependencies change
+#
+# Files needed to run setup.py
+# - README.md
+# - VERSION
+# - scripts/garage
+# - src/garage/__init__.py
+# - setup.py
+COPY --chown=$USER:$USER README.md $HOME/code/garage/README.md
+COPY --chown=$USER:$USER VERSION $HOME/code/garage/VERSION
+COPY --chown=$USER:$USER scripts/garage $HOME/code/garage/scripts/garage
+COPY --chown=$USER:$USER src/garage/__init__.py $HOME/code/garage/src/garage/__init__.py
+COPY --chown=$USER:$USER setup.py $HOME/code/garage/setup.py
+WORKDIR $HOME/code/garage
+
+ENTRYPOINT ["docker/entrypoint-headless.sh"]
+
+# Pre-build pre-commit env
+COPY --chown=$USER:$USER .pre-commit-config.yaml $HOME/code/garage
+RUN git init && \
+  pip install pre-commit && \
+  pre-commit install && \
+  pre-commit install-hooks
+
+# Install deps (but not the code)
+# Note that we don't delete the pip cache for dev builds
+RUN touch $HOME/.mujoco/mjkey.txt && \
+  pip install .[all,dev] && \
+  rm $HOME/.mujoco/mjkey.txt
+
+# Add code stub last (ensures code changes have the shortest builds)
+COPY --chown=$USER:$USER . $HOME/code/garage
+
+# Build and install the sdist
+RUN python3.5 setup.py sdist && \
+    cp dist/*.tar.gz dist/garage.tar.gz && \
+    pip install --upgrade --upgrade-strategy eager dist/garage.tar.gz[all,dev]
+
+RUN cd benchmarks && python3.5 setup.py sdist && \
+    cp dist/*.tar.gz dist/benchmarks.tar.gz && \
+    pip install --upgrade --upgrade-strategy eager dist/benchmarks.tar.gz
+
+# Nvidia target #######################################################################################################
```

Comment:
If you're going to use fences, please make them 80 chars wide.

Reply:
Okay!